### PR TITLE
Repro for Issue #1480 - Could not load type 'GrainInterfaces.DerivedClassWithGenericConstraints'

### PR DESCRIPTION
**Pull request by peter-perot**
Repro test for https://github.com/dotnet/orleans/issues/1480

(Could not load type 'GrainInterfaces.DerivedClassWithGenericConstraints')

### CODE REVIEW

File: src/Orleans.sln [hunk 1]
```diff
@@ -188,6 +188,10 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "TestInterfaces", "TestInter
 EndProject
 Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "TestFSharpInterfaces", "TestFSharpInterfaces\TestFSharpInterfaces.fsproj", "{2375D7D5-9B30-493F-9C2E-A6B2811A3C5A}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "TestInternalDtos", "TestInternalDtos\TestInternalDtos.csproj", "{D2C5F0E3-E9C3-46F5-82E3-26D221246F59}"
```
Comment (sergeybykov) [inline src/Orleans.sln]: Why add two new projects? Can't this fit into the existing test projects?

Comment (peter-perot) [inline src/Orleans.sln]: @sergeybykov: I know this is annoying. The problem is that the issue only occurs when I use a DTO which is in a *separate* assembly, the DTO is marked as *serializable* and the DTO is in an assembly which *references* a type from the Orleans.dll. When the DTO is in the test grain interfaces assembly, the issue does not occur.

Maybe you find a way the get rid off the additional assemblies; currently I am not able to reconstruct the issue with at least the additional assembly which is referencing Orleans.dll.

**Comment by sergeybykov** [comment]
What if we put it into the `TestInterfaces` project and make it reference `Orleans.dll`? It currently only has `ICSharpBaseInterface` in it which needs to be in a separate assembly from where it is used but shouldn't care if the project references `Orleans.dll` or not.

**Comment by peter-perot** [comment]
@sergeybykov: I tried what you told me (putting it onto `TestInterfaces` and letting `TestInterfaces` reference `Orleans.dll`), but then the test passes. But then again, if I use my own Assembly, the unit tests runs into a time-out exception.

File: test/TestInternalDtosRefOrleans/ClassReferencingOrleansTypeDto.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using Orleans;
+
+namespace UnitTests.DtosRefOrleans
+{
+    [Serializable]
+    public class ClassReferencingOrleansTypeDto
+    {
+        static ClassReferencingOrleansTypeDto()
```
Comment (sergeybykov) [inline test/TestInternalDtosRefOrleans/ClassReferencingOrleansTypeDto.cs]: It looks like this static constructor is breaking the test. If I remove it or make it non-static, the test passes fine.

I'm not convinced this is an Orleans issue. Static constructors with cross-assembly dependencies are dangerous due to non-deterministic order of their execution.

Comment (sergeybykov) [inline test/TestInternalDtosRefOrleans/ClassReferencingOrleansTypeDto.cs]: Never mind. The test is failing even with non-static constructor.

**Comment by sergeybykov** [comment]
@peter-perot Can you rebase this, so that aligns with the new location of test projects? If you are not sure how to do it, we can help with that.

**Comment by peter-perot** [comment]
@sergeybykov: I think I did a rebase from Orleans master into my branch since the test projects are correctly located in "orleans/test" (and not in "orleans/src/...). Did I miss something?

**Comment by sergeybykov** [comment]
@peter-perot  Sorry, you are right, of course. I just submitted https://github.com/peter-perot/orleans/pull/2 that removes one of the projects and marks the failing test as `SkippableFact`. If you merge it and CI is happy, then I think we can merge this one.

**Comment by peter-perot** [comment]
@sergeybykov: Done. But I don't see that CI has been triggered as it was done before. How can I kick CI to process my merge?

**Comment by sergeybykov** [comment]
@dotnet-bot test this please

**Comment by sergeybykov** [comment]
@peter-perot I saw that you merged my PR but I don't see this line in your commit - https://github.com/sergeybykov/orleans/commit/872357ef8c8ecd9c2e70add9603e9a971b9e6f6f#diff-571a4db393cbe49a4a6aeac2525060d8R23. Did you lose it while squashing?

File: test/TesterInternal/CodeGeneratorTests_TakeSerializedData.cs [hunk 3]
```diff
@@ -20,7 +20,7 @@ public class CodeGeneratorTests_TakeSerializedData : HostedTestClusterEnsureDefa
             await grain.TakeSerializedData(new Dtos.ClassNotReferencingOrleansTypeDto { MyProperty = "Test" });
         }
 
-        [Fact, TestCategory("BVT"), TestCategory("Functional"), TestCategory("CodeGen"), TestCategory("Serialization")]
+        [SkippableFact, TestCategory("BVT"), TestCategory("Functional"), TestCategory("CodeGen"), TestCategory("Serialization")]
```
Comment (peter-perot) [inline test/TesterInternal/CodeGeneratorTests_TakeSerializedData.cs]: @sergeybykov: I simply made an automerge via github webclient. And after fetching the branch, everything looks fine in my workspace: [Fact] has been replaced by [SkippableFact].

Sorry, I'm just trying to get familiar with the github stuff – maybe I'm missing something.

**Comment by peter-perot** [comment]
@dotnet-bot: test this please.

**Comment by sergeybykov** [comment]
@peter-perot Something is wrong here. You have both src/TesterInternal/CodeGeneratorTests_TakeSerializedData.cs and test/TesterInternal/CodeGeneratorTests_TakeSerializedData.cs. There are no src/Test+ folders/projects in master.

Let me know if I can help with fixing your branch.

**Comment by peter-perot** [comment]
@sergeybykov: Thanks, I see. The folder "src/TesterInternal/" was kind of orphaned, since it was not in the project anymore – which is okay so far, since we have a relocated version in "test\" (as you said).

I removed "src/TesterInternal\". Now my branch "Issue_1480" should be okay.

**Comment by peter-perot** [comment]
Question: The last two builds concerning this PR Jenkins complained that the "SkippableFact" failed. But I thought that test methods marked as skippable would pass the CI nevertheless?

**Comment by peter-perot** [comment]
Update: CI failed again on the [SkippableFact]. Don't know what's wrong. However – feel free, @sergeybykov, to adjust my code if there is still something wrong.

**Comment by sergeybykov** [comment]
@peter-perot Sorry, I mislead you -  `[SkippableFact]` skips the test conditionally, when a `SkipException` is thrown. Instead, change the attribute to `[Fact(Skip = "reproduces issue #1480"), TestCategory...`. That should make CI happy.

**Comment by peter-perot** [comment]
@dotnet-bot: test this please.

**Comment by peter-perot** [comment]
@sergeybykov: Yeah, I understand. Just changed the attribute, but now there seems to be a problem with Jenkins (see JVM stack trace). I think you can fix that.

**Comment by sergeybykov** [comment]
@dotnet-bot test this please

**Comment by sergeybykov** [comment]
Looks like it was a Jenkins fluke. 

@peter-perot Can you now squash 9 commits into 1 and rebase before we merge this?

**Comment by sergeybykov** [comment]
@peter-perot I rebased and squashed to https://github.com/sergeybykov/orleans/tree/1480_rebased. If it's too much trouble for you to fix your current branch, simply pull to a new branch off current master from https://github.com/sergeybykov/orleans/tree/1480_rebased, and resubmit the PR.

**Comment by sergeybykov** [comment]
@peter-perot Would you like me to submit a PR for https://github.com/sergeybykov/orleans/tree/1480_rebased instead?

**Comment by peter-perot** [comment]
@sergeybykov: I'll be back after the weekend and have a look at the PR stuff.

**Comment by sergeybykov** [comment]
@peter-perot No problem. Let us know then. It's ready to be merged otherwise.

**Comment by sergeybykov** [comment]
Replaced by https://github.com/dotnet/orleans/pull/1658.

**Pull request closed without merging**